File: backend/app/schemas/paper_editor.py
from datetime import datetime

from pydantic import BaseModel


class PaperProjectSummary(BaseModel):
    project_name: str
    title: str
    updated_at: datetime
    pdf_exists: bool
    pdf_url: str | None = None
    tex_url: str | None = None


class PaperFileRead(BaseModel):
    project_name: str
    file_path: str
    content: str
    updated_at: datetime


class PaperFileUpdate(BaseModel):
    content: str


class PaperCompileResponse(BaseModel):
    project_name: str
    success: bool
    pdf_url: str | None = None
    log: str
    compiled_at: datetime

File: frontend/src/services/paperEditor.ts
import { api } from './api'

export interface PaperProjectSummary {
  project_name: string
  title: string
  updated_at: string
  pdf_exists: boolean
  pdf_url?: string | null
  tex_url?: string | null
}

export interface PaperFileRead {
  project_name: string
  file_path: string
  content: string
  updated_at: string
}

export interface PaperCompileResponse {
  project_name: string
  success: boolean
  pdf_url?: string | null
  log: string
  compiled_at: string
}

export async function fetchPaperProjects() {
  const { data } = await api.get<PaperProjectSummary[]>('/paper-editor/projects')
  return data
}

export async function fetchProjectFiles(projectName: string) {
  const { data } = await api.get<string[]>(`/paper-editor/projects/${projectName}/files`)
  return data
}

export async function fetchProjectFile(projectName: string, filePath: string) {
  const { data } = await api.get<PaperFileRead>(
    `/paper-editor/projects/${projectName}/files/${filePath}`,
  )
  return data
}

export async function saveProjectFile(projectName: string, filePath: string, content: string) {
  const { data } = await api.put<PaperFileRead>(
    `/paper-editor/projects/${projectName}/files/${filePath}`,
    { content },
  )
  return data
}

export async function compileProject(projectName: string) {
  const { data } = await api.post<PaperCompileResponse>(
    `/paper-editor/projects/${projectName}/compile`,
  )
  return data
}
